### DISCUSSION TITLE
Does phase4 only can handle xml file?

### DISCUSSION THREAD

**Discussion by TMain-Tomi** [phax/phase4#95] [Q&A] [answered]
Hi Philip,

I find that the test cases from phase4-peppol-client only handle xml files.  If I want to handle other format, how can I do?

Best Regards,
Tomi Lin

**Comment by phax** [comment, marked as answer]
Hi @MrLinZHui 
phase4 is a generic solution, that can submit any kind of payload, not just XML.

`phase4-peppol-client` is a specific client that is tailored to the [Peppol](www.peppol.org) project which mandates the usage of "StandardBusinessDocumentHeader" (SBDH) which is an XML format. That's why the peppol-cient is bound to XML.

Use `Phase4Sender.builderUserMessage` as a starting point for configuring sending an "arbitrary" message. It uses the builder pattern to configured the message to be send.
Use e.g. `.payload (Phase4OutgoingAttachment.builder ().data (aPayloadBytes).mimeTypeXML ())` to provide arbitrary payload with arbitrary MIME types.

hth

**Reply by TMain-Tomi** [reply]
Dear Philip,

According to your guidance，I modified the code as follow:
/**
      _eResult = Phase4Sender.builderUserMessage()
              .fromPartyID("test")
              .fromPartyIDType("test")
              .fromRole("code>null</code>")
              .toPartyID("phase4")
              .toRole("<code>null</code>")
              .toPartyIDType("test")
              .action("<code>null</code>")
              .service("<code>null</code>")
              .addMessageProperty(ebms3Property)
              .addMessageProperty(property)
              .agreementRef("urn:fdc:peppol.eu:2017:agreements:tia:ap_provider")
              .payload(Phase4OutgoingAttachment.builder ().data (getBytesByFile(file)).mimeType(CMimeType.TEXT_PLAIN))
              .receiverCertificate(cert).endpointURL("http://localhost:8080/as4")
              .rawResponseConsumer(new AS4RawResponseConsumerWriteToFile ()).sendMessageAndCheckForReceipt();_

**/

And I started "phase4-spring-boot-demo" to receive msg, and got the error as follow:
/**
2022-08-20 14:08:54.251  INFO 24816 --- [nio-8080-exec-1] c.h.p.s.s.Phase4KeyStoreCallbackHandler  : Found keystore password for alias '1' and usage DECRYPT
2022-08-20 14:08:54.470 ERROR 24816 --- [nio-8080-exec-1] c.h.j.LoggingJAXBReadExceptionHandler    : **Failed to parse XML document: Content is not allowed in prolog.**
2022-08-20 14:08:54.485 ERROR 24816 --- [nio-8080-exec-1] s.Phase4PeppolServletMessageProcessorSPI : **[cdb4ca85-124f-438d-831f-378497cb46fc@phase4] Peppol SBDH Issue: [fatal_error] @ (1:1) Content is not allowed in prolog. (org.xml.sax.SAXParseException: Content is not allowed in prolog.)**
2022-08-20 14:08:54.485  WARN 24816 --- [nio-8080-exec-1] c.h.phase4.servlet.AS4RequestHandler     : Error invoking synchronous SPIs
**/

I can local the error when set "standardBusinessDocument" on "Phase4PeppolServletMessageProcessorSPI"
/**
        **a.m_aSBDH = SBDHReader.standardBusinessDocument ()
                              .setValidationEventHandler (new WrappedCollectingValidationEventHandler (aSBDHErrors))
                              .read (a.m_aPayloadBytes);**

**/ 
May you give me some advices? Thanks

Best Regards,
Tomi Lin

**Reply by phax** [reply]
Hi Tomi Lin,

I need a bit more background information on what you want to try to achieve. Usually AS4 is always used in a specific "profile" that defines common parts of the "PMode" (the parameters for point to point connections). Such profiles are "Peppol", "CEF", "ENTSOG" or "BPC" (mostly the entities publishing the profiles).
If you "just" want to use AS4 for some testing message exchange, you should define your own profile first.

If you are sending a "text/plain" content like your sending code suggests, then you cannot expect that the "Phase4PeppolServletMessageProcessorSPI" can handle the message correct. You need to create your own receiving SPI handler like the class `ExampleReceiveMessageProcessorSPI` in submodule `phase4-server-webapp`. In case you are not familiar with SPI, see https://docs.oracle.com/javase/tutorial/ext/basics/spi.html

**Reply by phax** [reply]
@MrLinZHui  did you progress on the issue? Do you have any additional information that would help in providing a more detailed response?

**Reply by TMain-Tomi** [reply]
Dear Philip,

I have solved this problem. Thank you for your guidance. 

Best Regards,
Tomi Lin

**Comment by TMain-Tomi** [comment]
Dear @phax ,

I meet a doubt when I used **AS4ClientUserMessage** to send a signed and encrypted message.
When I used private key to sign and send, it works successful.
When I used partner's public cert to encrypy and send, it works successful too.
But when I used private key and partner's public public to send a signed and encrypted message, it was failed.
I look for the test case **AS4ClientUserMessageTest** from **phase4-test** and found that it use same key alias(ph-as4).
/**
  private static AS4ClientUserMessage _setKeyStoreTestData (@Nonnull final AS4ClientUserMessage aClient)
  {
    final AS4CryptoProperties aCP = new AS4CryptoProperties ().setKeyStoreType (EKeyStoreType.JKS)
                                                              .setKeyStorePath ("keys/dummy-pw-test.jks")
                                                              .setKeyStorePassword ("test")
                                                              .setKeyAlias ("ph-as4")
                                                              .setKeyPassword ("test");
    aClient.setAS4CryptoFactory (new AS4CryptoFactoryProperties (aCP));
    aClient.cryptParams ().setAlias (aCP.getKeyAlias ());
    return aClient;
  }
**/
If I want to use **myself's private key** to sign and **partner's public cert** to encrypt, may you give me some advices?

Best Regards,
Tomi Lin

**Reply by phax** [reply]
The problem here is this line:
```
aClient.cryptParams ().setAlias (aCP.getKeyAlias ());
```
This line should use the Keystore alias of your partners public certificate. By passing `aCP.getKeyAlias ()` you say: use the same alias of my private key that I use for signing.

hth

**Reply by TMain-Tomi** [reply]
Dear @phax ,

Yes, it works successful now.
Thank you for your guidance.

Best Regards,
Tomi Lin
